Fix zero-dim handling in torch.matmul

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#63359 Fix zero-dim handling in torch.matmul**

Fixes #63352. The problem was that in e.g. `torch.matmul(A, B)` with A,
B having shapes [3, 2, 0] and [0, 2], the code attempts to call
`A.view(-1, 0)` which fails due to "-1 being ambiguous". The solution is
to manually compute what we want the shape of the view to be.

Test Plan:
- new tests

Differential Revision: [D30351583](https://our.internmc.facebook.com/intern/diff/D30351583)

zou3519 committed:
Fix zero-dim handling in torch.matmul

Fixes #63352. The problem was that in e.g. `torch.matmul(A, B)` with A,
B having shapes [3, 2, 0] and [0, 2], the code attempts to call
`A.view(-1, 0)` which fails due to "-1 being ambiguous". The solution is
to manually compute what we want the shape of the view to be.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30351583).

ngimel commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -1455,7 +1455,15 @@ Tensor matmul(
+    // This can happen in e.g. [3, 5, 0] @ [0, 0].
+    // So we manually compute the folding as a result.
+    auto dim1_size = 1;
+    for (const auto idx : c10::irange(0, size1.size() - 1)) {
```
use c10::multiply_integers instead of this loop

zou3519 replied:
Thanks for the pointer, I didn't know this existed.

lezcano reviewed (approved):
If this change adds instructions to matmul (I don't know if it does), it might be interesting to just check `if C10_UNLIKELY(numel() == 0)` at the beginning and handle this case separately.

zou3519 commented:
@Lezcano thanks for the suggestion to benchmark instruction counts. It looks like this PR adds one additional instruction to the non-zero-dim case https://gist.github.com/zou3519/e56d2619ddb3de5473c194f20f82c153 so it's not worth having a branch (the branch will add more instructions than it saves).

zou3519 committed:
Update on "Fix zero-dim handling in torch.matmul"


Fixes #63352. The problem was that in e.g. `torch.matmul(A, B)` with A,
B having shapes [3, 2, 0] and [0, 2], the code attempts to call
`A.view(-1, 0)` which fails due to "-1 being ambiguous". The solution is
to manually compute what we want the shape of the view to be.

Test Plan:
- new tests

Differential Revision: [D30351583](https://our.internmc.facebook.com/intern/diff/D30351583)

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30351583).
